internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+			// Obtain the information about the file currently at the lock file path.
+			var st unix.Stat_t
+			if err := unix.Stat(lockPath, &st); err != nil {
+				// The lock file we opened has been cleaned up by another process, which acquired
+				// and released the same lock file in between us releasing the lock at the start
+				// of this function, and then acquiring an exclusive lock again. There's nothing
+				// else for us to do.
+				return
+			}
+			if origSt.Ino != st.Ino {
+				// The lock file we opened has been cleaned up by another process, which acquired
+				// and released the same lock file in between us releasing the lock at the start
+				// of this function, and then acquiring an exclusive lock again. Another process
+				// has since created a new lock file. There's nothing else for us to do - the new
+				// process is responsible for cleaning up the new lock file.
+				return
+			}
+
+			// The lock file path still links to the lock file we originally opened and locked, and we
+			// have an exclusive lock on it again. As other processes participating in locking require
+			// an exclusive lock for cleaning it up, it os now safe to unlink it.
+			os.Remove(lockPath)
+		}, nil
+	}
+}
+
+// KDFType corresponds to a key derivation function.
+type KDFType string
+
+const (
+	KDFTypePBKDF2   KDFType = "pbkdf2"
+	KDFTypeArgon2i  KDFType = "argon2i"
+	KDFTypeArgon2id KDFType = "argon2id"
+)
+
+// Hash corresponds to a cryptpgraphic digest algorithm.
+type Hash string
+
+const (
+	HashSHA1   Hash = "sha1"
+	HashSHA224 Hash = "sha224"
+	HashSHA256 Hash = "sha256"
+	HashSHA384 Hash = "sha384"
+	HashSHA512 Hash = "sha512"
+)
+
+func (h Hash) GetHash() crypto.Hash {
+	switch h {
+	case HashSHA1:
+		return crypto.SHA1
+	case HashSHA224:
+		return crypto.SHA224
+	case HashSHA256:
+		return crypto.SHA256
+	case HashSHA384:
+		return crypto.SHA384
+	case HashSHA512:
+		return crypto.SHA512
+	default:
+		return 0
+	}
+}
+
+// KeyslotType corresponds to the type of a keyslot.
+type KeyslotType string
+
+const (
+	KeyslotTypeLUKS2 KeyslotType = "luks2"
+)
+
+// AFType corresponds to an anti-forensic splitter algorithm.
+type AFType string
+
+const (
+	AFTypeLUKS1 AFType = "luks1"
+)
+
+// AreaType corresponds to the type of a storage area in the binary
+// keyslots area.
+type AreaType string
+
+const (
+	AreaTypeRaw AreaType = "raw"
+)
+
+type label [48]byte
+
+func (l label) String() string {
+	return strings.TrimRight(string(l[:]), "\x00")
+}
+
+type csumAlg [32]byte
+
+func (a csumAlg) GetHash() crypto.Hash {
+	return Hash(strings.TrimRight(string(a[:]), "\x00")).GetHash()
+}
+
+type binaryHdr struct {
+	Magic       [6]byte
+	Version     uint16
+	HdrSize     uint64
+	SeqId       uint64
+	Label       label
+	CsumAlg     csumAlg
+	Salt        [64]byte
+	Uuid        [40]byte
+	Subsystem   [48]byte
+	HdrOffset   uint64
+	Padding     [184]byte
+	Csum        [64]byte
+	Padding4096 [7 * 512]byte
+}
+
+type jsonNumber string
```

Comment:
is this name used in the luks codebase? it's a bit confusing, if I'm not misreading this,  numbers are serialized as strings?

It's confusing also because there is json.Number that is something else. Unless there some external constraint I would call this luksNumber perhaps or something like that?

Reply:
I've renamed this to luksJsonNumber - ideally the code would just use json.Number but the LUKS2 format uses string-encoded uint64 numbers which aren't supported by json.Number.